feat(event-gateway): add guide for filtering records (#4095)
Kong/developer.konghq.com: app/_how-tos/event-gateway/filter-records-by-classification.md

    1. Create a Skip Records policy (consume phase) to filter records based on the header and principal name.

tools:
    - konnect-api

prereqs:
  inline:
    - title: Install kafkactl
      position: before
      include_content: knep/kafkactl
    - title: Define a context for kafkactl
      position: before
      include_content: knep/kafka-context
    - title: Start a local Kafka cluster
      position: before
      include_content: knep/docker-compose-start

cleanup:
  inline:
    - title: Clean up {{site.event_gateway}} resources
      include_content: cleanup/products/event-gateway
      icon_url: /assets/icons/gateway.svg

related_resources:
  - text: Modify Headers policy
    url: /event-gateway/policies/modify-headers/
  - text: Skip Records policy
    url: /event-gateway/policies/skip-record/
  - text: Schema Validation policy
    url: /event-gateway/policies/schema-validation-produce/
  - text: Expressions language
    url: /event-gateway/expressions/
---

## Overview

In this guide, you'll learn how to classify Kafka records at produce time and filter them at consume time based on user identity.

> We'll use a logging scenario where an `app_logs` topic contains log entries with different severity levels. Debug and trace logs are verbose and only useful for the SRE team troubleshooting issues, while regular developers only need info, warn, and error logs.

Reviewer: ```suggestion
We'll use a logging scenario where an `app_logs` topic contains log entries with different severity levels, aimed at two different groups of users:
* The SRE team needs debug and trace logs, which are verbose and useful for troubleshooting issues.
* Regular developers only need info, warn, and error logs.
```

The approach uses two policies:
1. **Produce phase**: A Schema Validation policy parses JSON records, and a nested Modify Headers policy adds an `x-internal: true` header to debug and trace logs.

> 2. **Consume phase**: A Skip Records policy filters out internal logs for users who aren't on the SRE team.

Reviewer: ```suggestion
2. **Consume phase**: A [Skip Records policy](/event-gateway/policies/skip-record/) filters out internal logs for users who aren't on the SRE team.
```

{% mermaid %}
flowchart LR
    P[Producer] --> SV

    subgraph produce [Event Gateway Produce policy chain]
        SV[Schema Validation<br/>Parse JSON] --> MH{Modify Headers<br/>level = debug/trace?}
        MH -->|Yes| H1[Add x-internal: true]
        MH -->|No| H2[No header added]
    end

    subgraph consume [Event Gateway Consume policy chain]
        SR{Skip Records<br/>x-internal = true<br/>AND user ≠ sre_user?}
        SR -->|Yes| DROP[Record skipped]
        SR -->|No| C[Send to consumer]
    end

    H1 --> K[Kafka Broker]
    H2 --> K
    K --> SR
    C --> CO[Consumer]
{% endmermaid %}

{:.info}
> **Performance tip**: Classifying records at produce time is more efficient than at consume time. Parsing JSON once during production avoids repeated deserialization for each consumer group.

## Create a Kafka topic

Create an `app_logs` topic in the Kafka cluster:

<!--vale off-->
{% validation custom-command %}
command: |
  kafkactl -C kafkactl.yaml --context direct create topic app_logs
expected:
  return_code: 0
render_output: false
{% endvalidation %}
<!--vale on-->

## Create a backend cluster

{% include knep/create-backend-cluster.md %}

## Create a virtual cluster

> Create a virtual cluster with two users:

Reviewer: ```suggestion
Create a virtual cluster with two users (`principals`):
```

{% endkonnect_api_request %}
<!--vale on-->

## Create a Skip Records policy to filter logs

Create a [Skip Records policy](/event-gateway/policies/skip-record/) that filters out internal logs for non-SRE users during the consume phase:

<!--vale off-->
{% konnect_api_request %}
url: /v1/event-gateways/$EVENT_GATEWAY_ID/virtual-clusters/$VIRTUAL_CLUSTER_ID/consume-policies
status_code: 201
method: POST
body:
  type: skip_record
  name: filter_internal_logs
  condition: "record.headers[\"x-internal\"] == \"true\" && context.auth.principal.name != \"sre_user\""
{% endkonnect_api_request %}
<!--vale on-->

## Configure kafkactl

Create a kafkactl configuration with contexts for both users:

<!--vale off-->
{% validation custom-command %}
command: |
  cat <<EOF > logs-cluster.yaml
  contexts:
    sre:
      brokers:
        - localhost:19092
      sasl:
        enabled: true
        username: sre_user
        password: sre_password
    dev:
      brokers:
        - localhost:19092
      sasl:
        enabled: true
        username: dev_user
        password: dev_password
  EOF
expected:
  return_code: 0
render_output: false
{% endvalidation %}
<!--vale on-->

## Validate

> 

Reviewer: ```suggestion
Now, let's validate that you can produce and consume log records as the two different user profiles.
```
We try to avoid having headings with no text, they get indexed poorly by search.
